lib/Dialect/TritonGPU/Transforms/WarpSpecialization/Partition.cpp

```diff
@@ -71,73 +49,82 @@ FailureOr<WarpSchedule> WarpSchedule::deserialize(scf::ForOp loop) {
   }
 
   for (Operation &op : loop.getBody()->without_terminator()) {
-    Partition *partition = result.getRootPartition();
-    if (auto attr = op.getAttrOfType<IntegerAttr>(kPartitionAttrName)) {
-      int64_t idx = attr.getInt();
-      if (idx < 0 || idx >= result.partitions.size())
-        return mlir::emitError(op.getLoc(), "invalid partition index ") << idx;
-      partition = result.partitions[idx].get();
+    if (auto attrs = getPartitionIds(&op)) {
+      for (auto idx : *attrs) {
+        if (idx < 0 || idx >= result.partitions.size())
+          return mlir::emitError(op.getLoc(), "invalid partition index ")
+                 << idx;
+        result.partitions[idx]->addOp(&op);
+      }
     }
-    result.insert(partition, &op);
   }
 
   return result;
 }
 
-void WarpSchedule::serialize(scf::ForOp loop) const {
-  SmallVector<Attribute> stages;
-  Builder b(loop.getContext());
-  for (Operation &op : loop.getBody()->without_terminator()) {
-    if (Partition *partition = opToPartition.lookup(&op)) {
-      if (partition == getRootPartition())
-        continue;
-      op.setAttr(kPartitionAttrName,
-                 b.getI32IntegerAttr(partition->getIndex()));
+void PartitionSet::dump() const {
+  for (auto [i, partition] :
+       llvm::enumerate(llvm::make_pointee_range(partitions))) {
+    llvm::errs() << "=== PARTITION #" << i << " ===\n";
+    for (Operation *op : partition.getOps()) {
+      op->print(llvm::errs(), OpPrintingFlags().skipRegions());
+      llvm::errs() << "\n";
     }
+    llvm::errs() << "\n";
   }
-  for (Partition &partition : getPartitions())
-    stages.push_back(b.getI32IntegerAttr(partition.getStage()));
-  loop->setAttr(kPartitionStagesAttrName, b.getArrayAttr(stages));
+  llvm::errs() << "\n";
 }
 
-LogicalResult WarpSchedule::verify(scf::ForOp loop) const {
-  // The root partition is only allowed to transitively depend on itself.
-  bool failed = false;
-  iterateInputs(loop, getRootPartition(), [&](OpOperand &input) {
-    auto [def, distance] = getDefiningOpAndDistance(loop, input.get());
-    // Ignore values defined outside the loop.
-    if (!def || def->getParentOp() != loop)
-      return;
-    const Partition *defPartition = opToPartition.at(def);
-    if (defPartition == getRootPartition())
-      return;
-    InFlightDiagnostic diag = mlir::emitWarning(input.getOwner()->getLoc());
-    diag << "operation in the root partition depends on a value that "
-            "originates from a non-root partition through operand #"
-         << input.getOperandNumber();
-    diag.attachNote(def->getLoc())
-        << "operand defined here in partition #" << defPartition->getIndex()
-        << " at distance " << distance;
-    failed = true;
-  });
-  if (failed)
-    return failure();
+namespace mlir::triton::gpu {
 
-  return success();
+void setPartition(Operation *op, ArrayRef<int> partitionIds) {
```

Comment:
odd ,it seem I am unable to use this setter with just one element, e.g.
`setPartition(op, {idx})` fails to compile, but this one compiles `setPartition(op, {idx,idx})`. Would like to be able to use the former as well.

Reply:
We should use the `setPartition(Operation *op, Partition *partition)` interface, which also adds the op to the op list of the given partition. `iterate*` functions walk such list for each partition.